src/Yesod/Auth/OAuth2/Dispatch.hs

```diff
@@ -100,7 +100,11 @@ withCallbackAndState
   -> Text
   -> m OAuth2
 withCallbackAndState name oauth2 csrf = do
-  uri <- ($ PluginR name ["callback"]) <$> getParentUrlRender
+  pluginURI <- ($ PluginR name ["callback"]) <$> getParentUrlRender
+  let
+    uri = case oauth2AppRoot oauth2 of
+      Just root -> root <> pluginURI
+      Nothing -> pluginURI
```

Comment:
Can you explain why you found this necessary?

If I understand it correctly, `getParentUrlRender` will respect [`Yesod(approot)`](https://hackage.haskell.org/package/yesod-core-1.6.24.4/docs/Yesod-Core.html#v:approot). Are you finding that you need to do to support a root here, but cannot use an overall [`Approot`](https://hackage.haskell.org/package/yesod-core-1.6.24.4/docs/Yesod-Core.html#t:Approot) value to accomplish it?

Reply:
The problem was that in the project I integrated this library with, the `approot` was set to `ApprootRelative` . Thus, the value generated by `getParentUrlRender` would have been a relative path, e.g. `/auth/plugin/okta/callback`.

Reply:
@pbrisbin Do you have any suggestions on how to better handle this case?

Reply:
There is an existing limitation that a project can't use a relative Approot and use this library -- therefore, this plugin can rely on it. I should probably dig up what that limitation was and properly document this :thinking:

Reply:
Oh it's "documented" here: https://github.com/freckle/yesod-auth-oauth2/blob/main/example/Main.hs#L56-L58

Reply:
@pbrisbin What do you think about support for Yesod projects with relative roots?

Reply:
Feel free to work on it, but we've been around a long time with this restriction in place and it's been OK. As far as I know OAuth2 (the protocol) needs to know domain for many reasons. So for use to lift our restriction on apps needing a static approot to use this library, we'd have to resort to some sort of "guess approot" function, which I think ultimately falls back on the `Host` header, which can be spoofed, which feels risky -- so I would need some convincing that it's safe to go that direction.

Reply:
@pbrisbin Could we change the behavior of `oauth2RedirectUri`? So, if the value is already set and valid we do no change it to the built `pluginURI` in the function `withCallbackAndState`.

Reply:
That way we can have support for relative appRoots without this hack.

Reply:
> Could we change the behavior of oauth2RedirectUri?

Hmm, I don't think so. That behavior (that field) comes directly from `hoauth2`:

![](https://files.pbrisbin.com/screenshots/screenshot.559726.png)

And its type (`URIRef Absolute`) is making it very clear, IMO, that it simply must be absolute. Therefore, to use OAuth2, you must use an absolute appRoot.

Reply:
I have created this PR to illustrate the change:
https://github.com/freckle/yesod-auth-oauth2/pull/178